blueprint/core/blueprint_EcmascriptEngine.cpp

```diff
@@ -207,27 +289,55 @@ namespace blueprint
         // Invocation
         if (duk_pcall(ctx, nargs) != DUK_EXEC_SUCCESS)
         {
-            if (onUncaughtError != nullptr)
-            {
-                const juce::String trace = duk_safe_to_stacktrace(ctx, -1);
-                const juce::String msg = duk_safe_to_string(ctx, -1);
+            errorHandler();
 
-                // Call the user provided error handler
-                std::invoke(onUncaughtError, msg, trace);
-
-                return juce::var::undefined();
-            }
-
-            duk_throw_raw(ctx);
+            // TODO: Return specific error code if invoke fails?
```

Comment:
@nick-thompson, would be great if you could just weigh in on the best error return approach for things like `EcmascriptEngine::invoke` and `EcmascriptEngine::evaluate`. I hear you on potentially using std::optional but personally I'm never a huge fan of std::optional as a return type if it can be avoided. It always feels like it bloats client code. We could just switch to a generic `EcmascriptEngineError`  type for all calls or alternatively provide an enum. I can see some possible benefits on using an enum and passing this to `onUncaughtError` callbacks so client code can handle specific error cases differently in future. 

Failing that though I'm not totally against std::optional, would just be good to have a consistent pattern.

Reply:
Yea, we should definitely figure this out. I like the std::optional idea but the hard part is that it's such an odd flow: "this might return something, and if it doesn't you should have already received a call to your uncaught error handler." Writing code like that sounds.. meh.

What do you think about this: `invoke` and `evaluate` can throw? If you try to evaluate something and it causes an error, we catch it in our own `uncaughtErrorHandler` and throw one of our own errors: either an `EcmascriptEngine::FatalError` or `EcmascriptEngine::Error`?

```cpp
try {
    auto result = engine.evaluate("2 + @");
    doSomethingWithResult(result); // Will never reach here
} catch (const EcmascriptEngine::Error& e) {
    // handle it
} catch (const EcmascriptEngine::FatalError& e) {
    // handle it
}
```

Writing try/catch code always feels a little verbose, but this feels like a good unifying story for how to interact with the engine, how to handle errors, how to differentiaite between regular and fatal errors, etc. What do you think?

Reply:
Yeah, it is all getting a bit "wordy". 

It still feels like everything should come through a single point. I guess try/catch is familiar to people. 

I wonder how bad it is to just have `onUncaughtError` take and error code for any special cases required and then have everything just return `juce::var::undefined` in the event of an error. 

For example, in `ReactApplicationRoot` we could call `evaluate` and have just `ReactApplicationRoot::handleError` responsible for setting `bundleIsValid`. Then we just return the `evaluate` result to the caller and don't check against any kinds of error codes. 

The only thing that "smells" to with the try/catch approach (as a client of the API) is "Wait .. what? I've written an `onUncaughtError` handler and I still need to stick try/catch all over the place?"

Reply:
Part of me still wonders if a single error code for returns and then all handling primarily through `onUncaughtError` makes sense as I'm talking this through. 

i.e. `static constexpr auto EcmascriptEngineError = "EcmascriptEngineError"`

The only other issue with the try/catch approach is callers can override `onUncaughtError` and take responsibility away from `ReactApplicationRoot`. Then it feels like it all gets a bit weird .... 

This could be solved by passing `onUncaughtError` as a constructor param to both `ReactApplicationRoot` and `EcmascriptEngine`

Reply:
Oh right, didn't explain that part of my thought process– with `try/catch` I think we could remove the whole idea of setting error callbacks as a user. We'll add our own error handling callbacks internally, but it will only ever surface to the user via try/catch. The reason I think that works nicely is that if `evaluate` always returns `juce::var::undefined()` in the case of an error, it's really hard to tell if you should proceed after the evaluate– you have to manage your own state to set a flag in the error handler callback and check it after every evaluate call. Although I don't love try/catch, at least it localizes the whole flow and relieves the end user from having to manage any state/flags in the error callback

Reply:
Cool. I'm all for this then in that case. I'll get to work on this now.

Reply:
@nick-thompson, just to throw one last curve ball. 

The other way I could see handling errors would be for all `EcmascriptEngine` functions to return a `juce::var`. We should only ever be calling the "safe" variants of the duktape API. In which case we can return an error code from all functions when we know errors are recoverable. For things like `EcmascriptEngine::registerNativeMethod` which don't have a real return result we can just return `juce::var::undefined`. The convention being that client code using the `EcmascriptEngine` API always checks as follows:

````
if (!EcmascriptEngine::thing() == EcmascriptEngine::Error)
```
Callers could just override a "fatal" error handler if needed. I.e. for the things we don't catch. 
It would then be reasonable enough for `ReactApplicationRoot` to keep it's `handleBundleError` function and maintain it's current behaviour. 

This is slightly less verbose than try/catch potentially? At the moment every public EcmascriptEngine API call could potentially throw. This could make for pretty cumbersome client code. What do you think?

Reply:
Hmmm. You know what. If `ReactApplicationRoot` is probably going to be used more often than anything not the exception use case might not be quite as bad... I'll wait to hear from you just in case. I can do a decent chunk of the refactor without needing to know anyway.

Reply:
I could imagine a convention like `if (!EcmascriptEngine::thing() == EcmascriptEngine::Error)` being similarly verbose to try/catch given that all the `evaluate`/`invoke` methods could yield a case that needs to be checked, so I think I still lean to try/catch. And you're right– most users won't interact with the EcmascriptEngine API directly, it will be through ReactAppRoot, where yea I like your idea that we only do the handling for them in Debug mode where we pop up the red UI screen (and can definitely add a flag to disable that). When the user moves to production they should probably try/catch their own call to `ReactAppRoot.evaluate` there

Reply:
Agreed, work for separate PR following this up I think. I have most of it in place.